lib/pg_rls.rb

```diff
@@ -66,32 +70,26 @@ def main_model
       class_name.to_s.camelize.constantize
     end
 
-    def on_each_tenant(&)
-      with_rls_connection do
-        result = []
-
-        main_model.find_each do |tenant|
-          allowed_search_fields = search_methods.map(&:to_s).intersection(main_model.column_names)
-          Tenant.switch tenant.send(allowed_search_fields.first)
-
-          result << { tenant:, result: ensure_block_execution(tenant, &) }
-        end
-
-        PgRls::Tenant.reset_rls!
-
-        result
-      end
+    def on_each_tenant(ids: [], scope: nil, &)
+      logger.deprecation_warning(
+        "PgRls.on_each_tenant is deprecated and will be removed in future versions. " \
+        "Please use PgRls::Tenant.on_find_each instead."
+      )
+      Tenant.on_find_each(ids: ids, scope: scope, &)
     end
 
     def execute_rls_in_shards
       connection_pool_list = PgRls.connection_class.connection_handler.connection_pool_list
       result = []
 
       connection_pool_list.each do |pool|
-        pool.connection.transaction do
-          Rails.logger.info("Executing in #{pool.connection.connection_class}")
+        connection = pool.lease_connection
```

Comment:
Perhaps another way to avoid the error is by adding a block like this:
```suggestion
        pool.with_connection do |connection|
          ...
        end
```